### PR TITLE
Merge from https://github.com/mohsen1/zipkin-api

### PR DESCRIPTION
**Pull request by mohsen1**

### CODE REVIEW

File: zipkin-api.yaml [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+  description: |
+    Zipkin's Query api is rooted at `api/v1`, on a host that by default listens
+    on port 9411. It primarily serves zipkin-web, although it includes a POST
+    endpoint that can receive spans.
+host: localhost:9411
+basePath: /api/v1
+schemes:
+  - http
+  - https
+consumes:
```
Comment (adriancole) [inline zipkin-api.yaml]: there's only one "consumes" aka POST endpoint. that also accepts application/x-thrift

File: zipkin-api.yaml [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+          description: Succes
+          schema:
+            type: array
+            items:
+              type: string
+        400:
+          description: Bad Request Error
+  /spans:
+    get:
+      parameters:
```
Comment (adriancole) [inline zipkin-api.yaml]: add description "Get all the span names logged by a particular service"

File: zipkin-api.yaml [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+      - name: serviceName
+        in: query
+        required: true
+        description: |
+          Ex zipkin-web (required) - service that logged an annotation in the
+          span.
+        type: string
+      responses:
+        200:
+          description: OK
```
Comment (adriancole) [inline zipkin-api.yaml]: ps this and /services are lexicographically sorted.. should this be somewhere? ex. here?

File: zipkin-api.yaml [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+      parameters:
+        - name: serviceName
+          in: query
+          required: true
+          description: service name
+          type: string
+        - name: annotationQuery
+          in: query
+          type: string
+          required: true
```
Comment (adriancole) [inline zipkin-api.yaml]: false

File: zipkin-api.yaml [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+        in: path
+        required: true
+        description: the 64-bit hex-encoded id of the trace as a path parameter.
+        type: string
+    get:
+      responses:
+        200:
+          description: OK
+          schema:
+            description: An array of spans (annotations?)
```
Comment (adriancole) [inline zipkin-api.yaml]: nope.. spans :) will look further down to see where the question came up

File: zipkin-api.yaml [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+            type: array
+            items:
+              $ref: "#/definitions/Annotation"
+        404:
+           description: "`traceId` not found"
+  /spans:
+    post:
+      description: Uploads a list of spans in json format
+      consumes:
```
Comment (adriancole) [inline zipkin-api.yaml]: this is the only consuming endpoint.. does that mean we should only declare consumes here?

File: zipkin-api.yaml [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+          schema:
+            description: An array of spans (annotations?)
+            type: array
+            items:
+              $ref: "#/definitions/Annotation"
+        404:
+           description: "`traceId` not found"
+  /spans:
+    post:
+      description: Uploads a list of spans in json format
```
Comment (adriancole) [inline zipkin-api.yaml]: a list of spans encoded per content-type

Spans may come in sparse, and be from different traces. For example apply may be called multiple times with a span with the same id, containing different annotations. These are merged by query time.

File: zipkin-api.yaml [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+      consumes:
+        - application/json
+        - application/x-thrift
+      parameters:
+        - name: span
+          in: body
+          description: A new span
```
Comment (adriancole) [inline zipkin-api.yaml]: A list of spans

File: zipkin-api.yaml [hunk 9]
```diff
@@ -0,0 +1,259 @@
...
+        type: string
+      ipv4:
+        type: string
+      port:
+        type: integer
+  TraceList:
+    type: array
+    items:
+      $ref: "#/definitions/Trace"
+  Trace:
```
Comment (adriancole) [inline zipkin-api.yaml]: this is a Span, not a Trace

File: zipkin-api.yaml [hunk 10]
```diff
@@ -0,0 +1,259 @@
...
+    type: object
+    properties:
+      traceId:
+        type: string
+      name:
+        type: string
+      id:
+        type: string
+      timestamp:
+        type: integer
```
Comment (adriancole) [inline zipkin-api.yaml]: does integer have a range? ex this is 64bit

File: zipkin-api.yaml [hunk 11]
```diff
@@ -0,0 +1,259 @@
...
+    title: Endpoint
+    type: object
+    properties:
+      serviceName:
+        type: string
+      ipv4:
+        type: string
+      port:
+        type: integer
+  TraceList:
```
Comment (adriancole) [inline zipkin-api.yaml]: ListOfSpans (for upload) or Trace (for query). Difference:

ListOfSpans - possibly different trace ids, in no particular order
Trace - List of spans who have the same trace id... Spans in the trace, and annotations in a span are sorted ascending by timestamp. ie first event should be first in the spans list.

**Comment by adriancole** [comment]
looks good.. only a few notes.

Next time, I'd raise a PR off a branch on your fork that isn't master. That way you can rebase your commit after feedback. http://rebaseandsqua.sh/

**Comment by karlunho** [comment]
Looks good to me too. Just tried it in apistudio

File: zipkin-api.yaml (hunk 1 above)
Comment (mohsen1) [inline zipkin-api.yaml]: That one is in here https://github.com/mohsen1/zipkin-api-1/commit/35a5531dc635a3aa9f4f943b6f080d23a053b08f#diff-bee5d8ba70d2a6b769696a3afe4341ccR141

The rest is JSON in, JSON out right?

File: zipkin-api.yaml (hunk 3 above)
Comment (mohsen1) [inline zipkin-api.yaml]: I'm not sure if you can describe sorting in JSON schema.

File: zipkin-api.yaml (hunk 6 above)
Comment (mohsen1) [inline zipkin-api.yaml]: Not sure about this. Till now I was thinking even empty body response considered plain text or if it's a JSON API an `application/json`. We can put `produces: []` to say this endpoint produces nothing though. I'm not sure how the rest of Swagger tooling thinks about it.

File: zipkin-api.yaml (hunk 10 above)
Comment (mohsen1) [inline zipkin-api.yaml]: We can use `format`

**Comment by mohsen1** [comment]
@adriancole I'll squash when you said it's all good. I totally agree. I should've created a different branch. I was just lazy ;)

**Comment by mohsen1** [comment]
Oh BTW I just pushed another commit that hopefully fixed all the issues you addressed.

File: zipkin-api.yaml (hunk 1 above)
Comment (adriancole) [inline zipkin-api.yaml]: there's no other place except for POST where json is "in", but yeah all out are json

File: zipkin-api.yaml (hunk 3 above)
Comment (adriancole) [inline zipkin-api.yaml]: np

File: zipkin-api.yaml [hunk 12]
```diff
@@ -0,0 +1,272 @@
...
+            type: array
+            title: ListOfDependencyLinks
+            items:
+              $ref: "#/definitions/DependencyLink"
+definitions:
+  Endpoint:
+    type: object
+    title: Endpoint
+    properties:
+      serviceName:
```
Comment (adriancole) [inline zipkin-api.yaml]: I assume I can add description keys to these later, right?

File: zipkin-api.yaml [hunk 13]
```diff
@@ -0,0 +1,272 @@
...
+          in: query
+          default: 10
+          type: integer
+          description: |
+            Maximum number of traces to return. Defaults to 10
+      responses:
+        200:
+          description: OK
+          schema:
+            $ref: "#/definitions/Trace"
```
Comment (adriancole) [inline zipkin-api.yaml]: response of the query is a ListOfTraces

File: zipkin-api.yaml [hunk 14]
```diff
@@ -0,0 +1,272 @@
...
+        type: string
+    get:
+      responses:
+        200:
+          description: OK
+          schema:
+            description: An array of spans
+            type: array
+            items:
+              $ref: "#/definitions/Annotation"
```
Comment (adriancole) [inline zipkin-api.yaml]: response is a Trace (aka ListOfSpans where traceId is common)

File: zipkin-api.yaml [hunk 15]
```diff
@@ -0,0 +1,272 @@
...
+    title: BinaryAnnotation
+    type: object
+    properties:
+      key:
+        type: string
+      value:
+        type: string
+      endpoint:
+        $ref: '#/definitions/Endpoint'
+  Endpoint:
```
Comment (adriancole) [inline zipkin-api.yaml]: repeat

File: zipkin-api.yaml [hunk 16]
```diff
@@ -0,0 +1,272 @@
...
+        - application/json
+        - application/x-thrift
+      produces: []
+      parameters:
+        - name: span
+          in: body
+          description: A list of spans
+          required: true
+          schema:
+            $ref: "#/definitions/ListOfSpans"
```
Comment (adriancole) [inline zipkin-api.yaml]: great

File: zipkin-api.yaml [hunk 17]
```diff
@@ -0,0 +1,272 @@
...
+        type: string
+      port:
+        type: integer
+  ListOfSpans:
+    title: ListOfSpans
+    description: 'A list of spans with possibly different trace ids, in no particular order'
+    type: array
+    items:
+      $ref: "#/definitions/Span"
+  Trace:
```
Comment (adriancole) [inline zipkin-api.yaml]: good

File: zipkin-api.yaml (hunk 10 above)
Comment (adriancole) [inline zipkin-api.yaml]: cool, can you add that to all timestamps? (assuming this is unsigned)

**Comment by adriancole** [comment]
couple bugs, then ready to go! I assume once this is merged, there's something to add to travis to validate and make a human readable doc page out of this?

File: zipkin-api.yaml (hunk 12 above)
Comment (mohsen1) [inline zipkin-api.yaml]: Yes of course

**Comment by mohsen1** [comment]
Fixed those issues. We can work on a simple validator as a TravisCI task

**Comment by adriancole** [comment]
@mohsen1 great! I'm going to cherry-pick this in, to clean up the commit.

**Comment by mohsen1** [comment]
I can squash as well

**Comment by mohsen1** [comment]
I just squashed into one commit

**Comment by adriancole** [comment]
too late :P

**Pull request closed without merging**